In [1]:
#Ejercicio 1: Limpieza de datos inicial
#Visualizacion del csv:

import pandas as pd

ruta = r'df_estados_bank.csv'
df = pd.read_csv(ruta, sep = ',')
df

Unnamed: 0  RowNumber  CustomerId  CreditScore  Gender    Balance  \
0              0          1    15634602          619  Female       0.00   
1              1          2    15647311          608  Female   83807.86   
2              2          3    15619304          502  Female  159660.80   
3              3          4    15701354          699  Female       0.00   
4              4          5    15737888          850  Female  125510.82   
...          ...        ...         ...          ...     ...        ...   
9995        9995       9996    15606229          771    Male       0.00   
9996        9996       9997    15569892          516    Male   57369.61   
9997        9997       9998    15584532          709  Female       0.00   
9998        9998       9999    15682355          772    Male   75075.31   
9999        9999      10000    15628319          792  Female  130142.79   

      NumOfProducts  HasCrCard  IsActiveMember   Age   Geography  \
0                 1          1               1  42.0       Texas   
1                 1          0               1  41.0  California   
2                 3          1               0  42.0       Texas   
3                 2          0               0  39.0       Texas   
4                 1          1               1  43.0  California   
...             ...        ...             ...   ...         ...   
9995              2          1               0  39.0       Texas   
9996              1          1               1  35.0       Texas   
9997              1          0               1   NaN       Texas   
9998              2          1               0  42.0     Alabama   
9999              1          1               0  28.0       Texas   

      EstimatedSalary    Surname  Exited  Tenure  
0           101348.88   Hargrave       1       2  
1           112542.58       Hill       0       1  
2           113931.57       Onio       1       8  
3            93826.63       Boni       0       1  
4            79084.10   Mitchell       0       2  
...               ...        ...     ...     ...  
9995         96270.64   Obijiaku       0       5  
9996        101699.77  Johnstone       0      10  
9997         42085.58        Liu       1       7  
9998         92888.52  Sabbatini       1       3  
9999         38190.78     Walker       0       4  

[10000 rows x 15 columns]

In [2]:
#No entiendo el resultado, deberia de darme descripcion del df y aqui veo el df completo
df.describe

<bound method NDFrame.describe of       Unnamed: 0  RowNumber  CustomerId  CreditScore  Gender    Balance  \
0              0          1    15634602          619  Female       0.00   
1              1          2    15647311          608  Female   83807.86   
2              2          3    15619304          502  Female  159660.80   
3              3          4    15701354          699  Female       0.00   
4              4          5    15737888          850  Female  125510.82   
...          ...        ...         ...          ...     ...        ...   
9995        9995       9996    15606229          771    Male       0.00   
9996        9996       9997    15569892          516    Male   57369.61   
9997        9997       9998    15584532          709  Female       0.00   
9998        9998       9999    15682355          772    Male   75075.31   
9999        9999      10000    15628319          792  Female  130142.79   

      NumOfProducts  HasCrCard  IsActiveMember   Age   Geography 

In [8]:
#LIMPIEZA DE DATOS INICIAL 
missing_values = df.isnull().sum()
missing_values

Unnamed: 0           0
RowNumber            0
CustomerId           0
CreditScore          0
Gender               0
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
Age                640
Geography            0
EstimatedSalary    792
Surname            294
Exited               0
Tenure               0
dtype: int64

*PREGUNTAS PARA DISCUTIR EN EQUIPO:*
¿QUE CONCLUSION PODEMOS SACAR QUE 'AGE' TENGA 640 DATOS NULOS, ASI COMO 'EstimatedSalary' (792) Y 'Surname'(294)?

¿SON IMPORTANTES? O LOS PODEMOS ELIMINAR?

In [11]:
#Eliminando filas duplicadas
df = df.drop_duplicates()
df

# ----> EL DF SE MANTUVO IGUAL, POR LO TANTO, NO HABIAN FILAS DUPLICADAS

Unnamed: 0  RowNumber  CustomerId  CreditScore  Gender    Balance  \
0              0          1    15634602          619  Female       0.00   
1              1          2    15647311          608  Female   83807.86   
2              2          3    15619304          502  Female  159660.80   
3              3          4    15701354          699  Female       0.00   
4              4          5    15737888          850  Female  125510.82   
...          ...        ...         ...          ...     ...        ...   
9995        9995       9996    15606229          771    Male       0.00   
9996        9996       9997    15569892          516    Male   57369.61   
9997        9997       9998    15584532          709  Female       0.00   
9998        9998       9999    15682355          772    Male   75075.31   
9999        9999      10000    15628319          792  Female  130142.79   

      NumOfProducts  HasCrCard  IsActiveMember   Age   Geography  \
0                 1          1               1  42.0       Texas   
1                 1          0               1  41.0  California   
2                 3          1               0  42.0       Texas   
3                 2          0               0  39.0       Texas   
4                 1          1               1  43.0  California   
...             ...        ...             ...   ...         ...   
9995              2          1               0  39.0       Texas   
9996              1          1               1  35.0       Texas   
9997              1          0               1   NaN       Texas   
9998              2          1               0  42.0     Alabama   
9999              1          1               0  28.0       Texas   

      EstimatedSalary    Surname  Exited  Tenure  
0           101348.88   Hargrave       1       2  
1           112542.58       Hill       0       1  
2           113931.57       Onio       1       8  
3            93826.63       Boni       0       1  
4            79084.10   Mitchell       0       2  
...               ...        ...     ...     ...  
9995         96270.64   Obijiaku       0       5  
9996        101699.77  Johnstone       0      10  
9997         42085.58        Liu       1       7  
9998         92888.52  Sabbatini       1       3  
9999         38190.78     Walker       0       4  

[10000 rows x 15 columns]

In [ ]:
#VERIFICAR, PORQUE LOS DATOS YA ERAN ENTEROS, ¿ERA NECESARIO?
# Convertimos las columnas al tipo de datos adecuado
df['HasCrCard'] = df['HasCrCard'].astype('int')
df['IsActiveMember'] = df['IsActiveMember'].astype('int')
df['Exited'] = df['Exited'].astype('int')

In [12]:
# Información general del DataFrame después de la limpieza
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       10000 non-null  int64  
 1   RowNumber        10000 non-null  int64  
 2   CustomerId       10000 non-null  int64  
 3   CreditScore      10000 non-null  int64  
 4   Gender           10000 non-null  object 
 5   Balance          10000 non-null  float64
 6   NumOfProducts    10000 non-null  int64  
 7   HasCrCard        10000 non-null  int64  
 8   IsActiveMember   10000 non-null  int64  
 9   Age              9360 non-null   float64
 10  Geography        10000 non-null  object 
 11  EstimatedSalary  9208 non-null   float64
 12  Surname          9706 non-null   object 
 13  Exited           10000 non-null  int64  
 14  Tenure           10000 non-null  int64  
dtypes: float64(3), int64(9), object(3)
memory usage: 1.1+ MB


INTERPRETARLO EN EQUIPO
CREO QUE SERÍA BUENO APLICAR UN 'df.info()' al inicio para comparar

1) PREGUNTA DEL RETO: ¿Consideráis que todas las variables del conjunto de datos son útiles de cara a la predicción? ¿Cuáles sí y cuáles no? 

<RowNumber: Es un índice del registro y no aporta información útil para la predicción. Puede ser eliminado.

<CustomerId: Es un identificador único de cliente y no aporta información predictiva directa. Puede ser eliminado.

<Surname: El apellido del cliente no tiene relevancia en la predicción del churn. Puede ser eliminado.

>CreditScore: La puntuación de crédito del cliente puede ser un predictor importante.

>Geography: El estado del cliente puede influir en su comportamiento bancario.

>Gender: El género del cliente puede tener algún impacto en el comportamiento.

>Age: La edad del cliente es un factor relevante para el análisis.

>Tenure: El número de años que el cliente lleva en el banco es un factor importante.

>Balance: El balance del cliente puede ser crucial para predecir el churn. (AVERIGUAR QUE ES BALANCE)

>NumOfProducts: El número de productos que el cliente tiene en el banco puede influir en su lealtad.

>HasCrCard: Tener o no tarjeta de crédito puede ser relevante.

>IsActiveMember: Ser un cliente activo puede influir en la probabilidad de churn.

>EstimatedSalary: El salario estimado del cliente puede ser un factor importante.

>Exited: Es la variable objetivo que estamos tratando de predecir.

<Por lo tanto, las variables útiles para la predicción son:>

CreditScore
Geography
Gender
Age
Tenure
Balance
NumOfProducts
HasCrCard
IsActiveMember
EstimatedSalary



In [17]:
# Procedemos a eliminar columnas no útiles

df_cleaned = df.drop(columns=['RowNumber', 'CustomerId', 'Surname'])
df_cleaned.head()

Unnamed: 0  CreditScore  Gender    Balance  NumOfProducts  HasCrCard  \
0           0          619  Female       0.00              1          1   
1           1          608  Female   83807.86              1          0   
2           2          502  Female  159660.80              3          1   
3           3          699  Female       0.00              2          0   
4           4          850  Female  125510.82              1          1   

   IsActiveMember   Age   Geography  EstimatedSalary  Exited  Tenure  
0               1  42.0       Texas        101348.88       1       2  
1               1  41.0  California        112542.58       0       1  
2               0  42.0       Texas        113931.57       1       8  
3               0  39.0       Texas         93826.63       0       1  
4               1  43.0  California         79084.10       0       2

In [19]:
#2)Implementación de algoritmos predictivos
#Separamos los datos en conjuntos de entrenamiento y prueba
from sklearn.model_selection import train_test_split

# Definir variables predictoras (X) y objetivo (y)
X = df_cleaned.drop(columns=['Exited'])
y = df_cleaned['Exited']  # --> Variable Objetivo

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

<Estamos enfrentando un problema de clasificación binaria (predecir si un cliente abandona o no el banco), por lo que usaremos modelos de clasificación como:

-Regresión Logística
-Árboles de Decisión
-Random Forest
- Gradient Boosting ---------> (VERIFICAR SI LO HEMOS ESTUDIADO, NO LO RECUERDO)  ELIMINAR

In [23]:
#APLICAMOS LOS DIFERENTES ALGORITMOS SELECCIONADOS PARA COMPARAR CUAL ES EL QUE MAS NOS CONVIENE:

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Inicializar los modelos
log_reg = LogisticRegression()
tree_clf = DecisionTreeClassifier()
rf_clf = RandomForestClassifier()
gb_clf = GradientBoostingClassifier()

# Entrenar los modelos
log_reg.fit(X_train, y_train)
tree_clf.fit(X_train, y_train)
rf_clf.fit(X_train, y_train)
gb_clf.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred_log_reg = log_reg.predict(X_test)
y_pred_tree_clf = tree_clf.predict(X_test)
y_pred_rf_clf = rf_clf.predict(X_test)
y_pred_gb_clf = gb_clf.predict(X_test)

#Usamos métricas de evaluación para comparar los modelos:
def evaluate_model(y_test, y_pred):
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    return accuracy, precision, recall, f1

#Evaluamos cada modelo
metrics_log_reg = evaluate_model(y_test, y_pred_log_reg)
metrics_tree_clf = evaluate_model(y_test, y_pred_tree_clf)
metrics_rf_clf = evaluate_model(y_test, y_pred_rf_clf)
metrics_gb_clf = evaluate_model(y_test, y_pred_gb_clf)

# Resultados
print("Logistic Regression: ", metrics_log_reg)
print("Decision Tree: ", metrics_tree_clf)
print("Random Forest: ", metrics_rf_clf)
print("Gradient Boosting: ", metrics_gb_clf)

ValueError: could not convert string to float: 'Female'

*ATENCION*
DEBEMOS CONVERTIR LAS VARIABLES CATEGÓRICAS EN VARIABLES NUMÉRICAS

AHORA NOS ESTÁ DANDO PROBLEMAS CON EL "GENDER", PERO SOSPECHO QUE DESPUES NOS DARÁ PROBLEMAS TAMBIEN "GEOGRAPHY"